### DISCUSSION TITLE
Padavan и ошибка iptables: No chain/target/match by that name

### DISCUSSION THREAD

**Discussion by cybsuai** [Anonym-tsk/nfqws-keenetic#23] [Делимся опытом]
Удалось с бубном выполнить установку на Mi R3G с прошивкой Padavan с Entware во внутренней памяти.
Но имею данную ошибку при старте. Для кинетика выход известен, а есть ли решение для Padavan?

admin@MI-R3G:/opt/home/admin$ /opt/etc/init.d/S51nfqws start
xt_multiport.ko loaded
xt_connbytes.ko loaded
xt_NFQUEUE.ko loaded
Loading hostlist /opt/etc/nfqws/user.list
loading plain text list
Loaded 16 hosts from /opt/etc/nfqws/user.list
iptables: No chain/target/match by that name.
iptables: No chain/target/match by that name.
iptables: No chain/target/match by that name.
Started NFQWS service

Увидел, что из загружаемых модулей есть лишь только xt_connbytes.ko, а xt_multiport.ko и xt_NFQUEUE.ko нет в принципе.
Что с этим делать, пока непонятно...

**Comment by Anonym-tsk** [comment]
Покажите вывод `iptables -t mangle -L`

**Reply by cybsuai** [reply]
Полночи просидел сегодня, сегодня уже собирался отписаться о победе :)
В общем основная проблема, в стандартной прошивке Padavan не хватает компонента NFQUEUE. Пересобрал прошивку прометеусом от hadzhioglu и его же репозитория, изменив в конфиге для своего девайса: 
# CONFIG_NETFILTER_XT_TARGET_NFQUEUE is not set
на
# CONFIG_NETFILTER_XT_TARGET_NFQUEUE=y
Тогда при условии некоторых модификаций вашего скрипта под условия padavana всё начало отлично работать!
Вывод теперь соответственно такой:

Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
NFQUEUE    tcp  --  anywhere             anywhere             tcp dpt:www connbytes 1:8 connbytes mode packets connbytes direction original mark match ! 0x40000000/0x40000000 NFQUEUE num 200 bypass
NFQUEUE    tcp  --  anywhere             anywhere             tcp dpt:https connbytes 1:8 connbytes mode packets connbytes direction original mark match ! 0x40000000/0x40000000 NFQUEUE num 200 bypass
NFQUEUE    udp  --  anywhere             anywhere             udp dpt:https connbytes 1:8 connbytes mode packets connbytes direction original mark match ! 0x40000000/0x40000000 NFQUEUE num 200 bypass
 
Короче, теперь всё работает, но боюсь что мало кто сможет это легко повторить.
Если интересны подробности что пришлось изменить и что в исходном варианте не работает на падаване, могу поделиться.

**Reply by GoldVital** [reply]
Поделитесь, пожалуйста!

**Reply by cybsuai** [reply]
Установочный скрипт если я правильно помню, проходит почти без ошибок - если заранее сделать "opkg install busybox". Скрипт не сможет скачать исполняемый файл для nfqws, но у меня он уже был - так что я не разбирался что там пошло не так.
Сам скрипт я немного изменил под реалии падавана, собранного с нужными пакетами - загружать требуется только один дополнительный модуль. Также добавил автоматическое определение интерфейса, куда навешивать правила и добавил ведение лога.
Чтобы всё корректно переприменялось, в вебинтерфейсе роутера Персонализация-Скрипты-Выполнить после перезапуска правил брандмауэра добавил строчку /opt/etc/init.d/S51nfqws restart

Сам скрипт
[S51nfqws.zip](https://github.com/user-attachments/files/16711741/S51nfqws.zip)

Cобирал прошивки с модулями xt_multiport, xt_connbytes и xt_NFQUEUE для роутеров Xiaomi R3G, Mini и Redmi Router AC2100. 
Бинарники nfqws для этих роутеров брать тут: https://github.com/bol-van/zapret/tree/master/binaries/mips32r1-lsb

**Reply by cybsuai** [reply]
Ой, ну и главное, чтобы добавить поддержку модулей в прошивку: менять в файле kernel-3.4.x.config для своего роутера
CONFIG_NETFILTER_XT_TARGET_NFQUEUE=y
CONFIG_NETFILTER_XT_MATCH_MULTIPORT=y

**Reply by GoldVital** [reply]
Спасибо! По логам вроде бы все встало нормально, понаблюдаю
![nfq](https://github.com/user-attachments/assets/9f106eb5-f7cf-40ee-888c-566538b449ed)

/opt/home/admin # iptables -t mangle -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination

**Reply by cybsuai** [reply]
Судя по логу, не встало. При установке как минимум видна ошибка iptables - не хватает модулей. 
А вывод  iptables -t mangle -L должен быть такой:
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
NFQUEUE    tcp  --  anywhere             anywhere             tcp dpt:www connbytes 1:8 connbytes mode packets connbytes direction original mark match ! 0x40000000/0x40000000 NFQUEUE num 200 bypass
NFQUEUE    tcp  --  anywhere             anywhere             tcp dpt:https connbytes 1:8 connbytes mode packets connbytes direction original mark match ! 0x40000000/0x40000000 NFQUEUE num 200 bypass
NFQUEUE    udp  --  anywhere             anywhere             udp dpt:https connbytes 1:8 connbytes mode packets connbytes direction original mark match ! 0x40000000/0x40000000 NFQUEUE num 200 bypass

То есть правила не встали. И наверное нет и самого процесса (у меня он тут 672): 
![image](https://github.com/user-attachments/assets/6ee502d4-cdba-4779-b37c-cfa8895668e5)

**Reply by GoldVital** [reply]
Да не стартует. Что можно подправить?
![st](https://github.com/user-attachments/assets/b8256230-7545-4790-b15c-2abe6c13c9a6)

**Reply by cybsuai** [reply]
Пока исходим из того, что вы собрали прошивку с правильными пакетами.
Тогда попробуйте заменить S51nfqws на тот, что я выкладывал чуть выше (только укажите в нём в начале путь к самой программе nfqws где она у вас лежит), и скопируйте сюда:
1) вывод на экран в момент старта /opt/etc/init.d/S51nfqws start
2) лог из /var/log/nfqws_log

**Reply by GoldVital** [reply]
/opt/home/admin # /opt/etc/init.d/S51nfqws start
/opt/etc/init.d/S51nfqws: line 58: /opt/home/admin/zapret/nfqws: not found
iptables: No chain/target/match by that name.
iptables: No chain/target/match by that name.
iptables: No chain/target/match by that name.
Started NFQWS service

Судя по вашему скрину и моему логу у меня не установлен zapret ))

**Reply by cybsuai** [reply]
Всё что у меня лежит в zapret:
![image](https://github.com/user-attachments/assets/36afa70f-1add-497f-85ad-7bba4636d77b)

А у вас проблема в том, что в прошивке нет нужных пакетов. Об этом говорит ошибка 
iptables: No chain/target/match by that name.
И даже если вы положите nfqws в zapret, он не стартанёт

Нужна прошивка с пакетами. У вас какой роутер?

**Reply by GoldVital** [reply]
у меня роутер Mi R3G

**Reply by cybsuai** [reply]
Для него могу поделиться готовой прошивкой, но завтра вечером - я собирал её на даче)))

**Reply by GoldVital** [reply]
> Для него могу поделиться готовой прошивкой, но завтра вечером - я собирал её на даче)))

Хорошо, подожду

**Reply by cybsuai** [reply]
> Хорошо, подожду

Вот она. Для R3G, не R3Gv2 или чего то ещё.
[MI-R3G_3.4.3.9L-102-c954b8f.zip](https://github.com/user-attachments/files/16730025/MI-R3G_3.4.3.9L-102-c954b8f.zip)

С ней точно работает моя версия скрипта, если положить в /opt/home/admin/zapret файл nfqws, а конфиги подходят те, что тут. В общем пишите что получится

**Reply by GoldVital** [reply]
> > Хорошо, подожду
> 
> Вот она. Для R3G, не R3Gv2 или чего то ещё. [MI-R3G_3.4.3.9L-102-c954b8f.zip](https://github.com/user-attachments/files/16730025/MI-R3G_3.4.3.9L-102-c954b8f.zip)
> 
> С ней точно работает моя версия скрипта, если положить в /opt/home/admin/zapret файл nfqws, а конфиги подходят те, что тут. В общем пишите что получится
![image](https://github.com/user-attachments/assets/aab16c6f-5aab-418d-a501-19309f654df1)
Огромное человеческое спасибо вам, теперь вроде работает

**Reply by Ghost-Solder** [reply]
Подскажите, пожалуйста, а как собирали? Хочу собрать под MI-R3P, сейчас прошивка padavan 3.4.3.9-099_20200507

**Reply by cybsuai** [reply]
> Подскажите, пожалуйста, а как собирали? Хочу собрать под MI-R3P, сейчас прошивка padavan 3.4.3.9-099_20200507

Воспользовался этим репозиторием как исходным https://4pda.to/forum/index.php?showtopic=714487&view=findpost&p=121299584
Далее скомпилировал прошивку для своего роутера, поменяв сначала в его конфиге /trunk/configs/boards/.../kernel-3.4.x.config две строчки на
CONFIG_NETFILTER_XT_TARGET_NFQUEUE=y
CONFIG_NETFILTER_XT_MATCH_MULTIPORT=y

**Reply by Ghost-Solder** [reply]
Спасибо, пробую

**Reply by cybsuai** [reply]
> Спасибо, пробую

Если не получится, скажите точно модель (MI-R3PRO или MI-R3P_SPI), скомпилирую

**Reply by Ghost-Solder** [reply]
Вообще на роутере указано (PRO) MI-R3P, SPI не перепаивал. Пока добрался до сюда
![image](https://github.com/user-attachments/assets/24f21dd1-9473-48b6-b9ee-41f02421caed)

**Reply by Ghost-Solder** [reply]
Сразу вопрос на будущее - чтобы прошить через веб-морду, скомпилированную прошивку нужно будет ещё под pb-boot подгонять?

**Reply by cybsuai** [reply]
Про это я не в курсе, думаю следует опираться на эту инструкцию (так как прошивка сделана из репозитория этого человека) https://4pda.to/forum/index.php?showtopic=810698&view=findpost&p=128045892

Сам я шил просто через вебморду с сохранением настроек и всё было хорошо.

ЗЫ вот под ваш дейвайс, если не получится самостоятельно
[MI-R3PRO_3.4.3.9L-102-c954b8f.zip](https://github.com/user-attachments/files/16825834/MI-R3PRO_3.4.3.9L-102-c954b8f.zip)

**Reply by Ghost-Solder** [reply]
Спасибо огромное! Как раз застрял на сборке тулчейна
![image](https://github.com/user-attachments/assets/5d593e53-a22d-4654-a607-4d6a661618f0)

**Reply by Ghost-Solder** [reply]
Что-то не взлетело - после прошивки подключения перестали работать, не удалось зайти в вебморду никаким из способов. Похоже на бутлуп. Пришлось откатиться через pb-boot на прошивку из этого коммента https://4pda.to/forum/index.php?showtopic=810698&view=findpost&p=96601661
Не знаете, а можно ли с помощью промитиуса пересобрать прошивку, которая уже работает, но с нужными пакетами?

**Reply by cybsuai** [reply]
Очень жаль, у вас я так понимаю не breed? 
В готовую прошивку думаю что уже ничего не внедрить, нужны исходники и сборка.

**Reply by Ghost-Solder** [reply]
Не breed. Полазил по 4pda и там действительно периодически выкладывали по две прошивки одну под pb-boot другую под breed (пример - https://4pda.to/forum/index.php?showtopic=810698&view=findpost&p=106601019). В чём отличие во время сборки, пока не понимаю

**Reply by lo1kek** [reply]
@cybsuai подскажи плиз в прошивке для R3G [MI-R3G_3.4.3.9L-102-c954b8f.zip](https://github.com/user-attachments/files/16730025/MI-R3G_3.4.3.9L-102-c954b8f.zip) файлы с модулями тут `/lib/modules/3.4.113`  должны быть?
Буквально вчера вечером всё поставил работало, а сегодня малость тюнил настройки nfqws и теперь ловлю `iptables: No chain/target/match by that name.` Файлов по пути выше нет.
Не могу вкурить как такое быть может, модули ядра же в read-only fs лежат, как они пропасть могли. Или их там и не было?
Прошивку со вчера не сносил, c954b8f в ui роутера.

**Reply by cybsuai** [reply]
Модулей в виде отдельного файла там не должно быть, кроме xt_connbytes. В моем скрипте как видите только он и грузится. 

Изначально я пробовал вариант с отдельными модулями, но не взлетело. В итоге они получилось включенными куда-то внутрь netfilter.

**Reply by lo1kek** [reply]
понял, тогда буду искать где косяк

**Reply by Ghost-Solder** [reply]
@cybsuai Спасибо большое! В итоге, в несколько заходов, по вечерам, удалось себе пересобрать прошивку под pb-boot с нужными модулями и с помощью вашего скрипта запустить. Даже не верится, что всё работает. Ещё раз спасибо!

PS Не сразу догадался, что кроме бинарника из оригинального запрета, нужно было перетащить ещё и папку nfqws из этой репы.

**Reply by cybsuai** [reply]
> Даже не верится, что всё работает.

Очень рад за вас! 
Если не секрет, расскажите как вы собрали прошивку для pb-boot, вдруг тоже пригодится. Я так и не нагуглил ничего полезного на эту тему

**Reply by Ghost-Solder** [reply]
> > Даже не верится, что всё работает.
> 
> Очень рад за вас! Если не секрет, расскажите как вы собрали прошивку для pb-boot, вдруг тоже пригодится. Я так и не нагуглил ничего полезного на эту тему

Конечно не секрет - вот на [4pda описание](https://4pda.to/forum/index.php?showtopic=810698&view=findpost&p=132150394). Коротко - просто, нашёл откуда была собрана моя текущая прошивка и собрал из той же репы, там как раз есть конфиг под breed и под pb-boot - https://github.com/hanwckf/rt-n56u

Наверняка, это не лучший способ, но как сделать лучше без загрузки breed'a не придумал.

**Reply by Anonym-tsk** [reply]
> Бинарники nfqws для этих роутеров брать тут: https://github.com/bol-van/zapret/tree/master/binaries/mips32r1-lsb

Это ж mipsel, зачем его перекачивать?

**Comment by piu-piu** [comment]
Здравствуйте. Роутер MI-R3G. Прошивка падаван. Проделал все те же шаги из диалога в этой ветке.

![kW9lKWgNrE](https://github.com/user-attachments/assets/c541f746-7d2c-4f67-9a1b-75abc3788cef)
![aR086OWY3L](https://github.com/user-attachments/assets/dff24523-4f6b-404d-b71c-5eaaf48af6f8)

Вроде никаких ошибок не наблюдается. Но все блокированное не работает. DOH и DOT включал\выключал как и Ipv6.
Подскажите, пожалуйста, куда еще смотреть?

**Comment by cybsuai** [comment]
Посмотрите, правильный ли у вас интерфейс указан, куда навешивать: 
route | grep '^default' | grep -o '[^ ]*$'

**Reply by piu-piu** [reply]
![aR086OWY3L](https://github.com/user-attachments/assets/c5fe9246-68d5-4a16-8211-641b7e089b5e)
Ну да, wan провайдерский.
![kW9lKWgNrE](https://github.com/user-attachments/assets/73076e19-ca5e-406e-a723-36bba78064d5)
При обращении к блокированным (твитер для примера, с другими аналогично) вот такая ошибка.
Подозреваю, что все же с настройками dns в роутере какая-то беда, но все возможные вариации вроде испробовал.

**Reply by cybsuai** [reply]
Тогда попробуйте ещё конкретно тот скрипт, что я выкладывал выше.

**Reply by piu-piu** [reply]
Сейчас именно бинарник [из запрета, ](https://github.com/bol-van/zapret/tree/master/binaries/mips32r1-lsb)
И [ваш скрипт](https://github.com/user-attachments/files/16711741/S51nfqws.zip)
На вашей же прошивке.

**Reply by cybsuai** [reply]
Понял, просто предыдущий ваш скрин был со скриптом из этого репозитория. 
А вы с перезагрузкой роутера пробовали? (Есть там нюанс с загрузкой библиотек)
Если да, у меня идеи кончились... демон работает, правила есть => должно работать.

**Reply by piu-piu** [reply]
Перезагрузка ничего не изменила.

**Reply by piu-piu** [reply]
Хм. Заново перекопировал ваш скрипт и бинарник и ютуб завелся. Но логику автолиста не могу разобрать. Домены добавляются, но не все открываются. NTC и еще один открылись, инста, фб, твитер, икс - нет. Ну и закрытые для рф темы 4пда тоже, конечно, нет. В общем для ТВ победил, а на пк продолжу пользоваться некобокс. Спасибо

**Comment by Mannokia** [comment]
Здравствуйте. Имею роутер Xiaomi Wi-Fi Router Redmi AC2100 и ту же проблему. Как добыть NetFilter для данного роутера?

**Reply by piu-piu** [reply]
Вам, скорее всего, пересобрать падаван нужно как это cybsuai сделал дли ми3рг.

**Reply by Mannokia** [reply]
Знать бы как это сделать ). Я пользуюсь чужой сборкой.

**Reply by piu-piu** [reply]
Prometheus на 4пда

**Reply by Mannokia** [reply]
Попробую конечно, но мне кажется что это будет для меня очень не просто.  Как туда добавлять этот модуль не ясно. Когда то давно я пользовался этим скриптом, но тогда на умолчании все компелировал.

**Reply by piu-piu** [reply]
Попробуйте задать вопрос в той теме. И еще тему своего устройства гляньте. Может уже все есть. А, вообще, там вроде помогают.

**Reply by cybsuai** [reply]
Ловите! Также есть прошивка с пакетами для Mi Mini, если кому надо спрашивайте. Других не собирал
[RM-AC2100_3.4.3.9L-102-c954b8f.zip](https://github.com/user-attachments/files/16825264/RM-AC2100_3.4.3.9L-102-c954b8f.zip)

**Reply by Mannokia** [reply]
> RM-AC2100_3.4.3.9L-102-c954b8f.zip

Спасибо
Но не помогло (((
![image](https://github.com/user-attachments/assets/cba9212f-555b-4f46-aa55-37be29c30ec6)

**Reply by cybsuai** [reply]
Так вам нужно скрипт взять тот, что я для падавана выкладывал... оригинальный работать и не должен

**Reply by Mannokia** [reply]
> Так вам нужно скрипт взять тот, что я для падавана выкладывал.

Прошу прощения. Я что то упустил. По пиреписке нашел скрипт, но не пойму как его  использовать.
[S51nfqws.zip](https://github.com/user-attachments/files/16825672/S51nfqws.zip)
Это он я полагаю.

**Reply by piu-piu** [reply]
https://github.com/Anonym-tsk/nfqws-keenetic/discussions/23#discussioncomment-10449678
замените бинарник и скрипт, в котором не забудьте заменить путь до бинарника

**Reply by cybsuai** [reply]
1) положить скрипт S51nfqws в /opt/etc/init.d
2) сделать исполняемым, если надо: chmod +x /opt/etc/init.d/S51nfqws
3) в скрипте (строка 10 вроде бы) изменить путь к nfqws, типа 
было: NFQWS_BIN=/opt/home/admin/zapret/nfqws
стало: NFQWS_BIN=/opt/usr/bin/nfqws

Перезагрузить роутер

**Reply by Mannokia** [reply]
Спасибо огромное. Все заработало. Даже без перезапуска роутера, просто   `/opt/etc/init.d/S51nfqws start` хватило

**Reply by Mannokia** [reply]
> замените бинарник

Какой бинарник следовало заменить?  Язаменил только скрипт S51nfqws и в нем поменял строки как написал  [cybsuai].
Боюсь что что то упустил. Поправьте меня пожалуйста.

**Reply by cybsuai** [reply]
Если уже работает, менять не надо ничего)

**Reply by GoldVital** [reply]
Добрый день! "Работает, не трож!" А под новую версию скрипта адаптация не планируется? Подкрутил , посмотрите, вроде даже работает
[S51nfqws.zip](https://github.com/user-attachments/files/16910350/S51nfqws.zip)

![IMG_20240907_150131](https://github.com/user-attachments/assets/8d347a8d-f8f6-4607-8348-83eab6d7e634)
![IMG_20240907_150117](https://github.com/user-attachments/assets/4d487893-a266-4014-a36b-ddb3120c37de)
![IMG_20240907_150105](https://github.com/user-attachments/assets/3ed197d0-ec4b-4165-a687-749b93bd0b7e)

**Comment by Anonym-tsk** [comment]
@cybsuai я так понимаю, основное отличие вашего скрипта в расположении модулей ядра?
```
/lib/modules/$KERNEL/kernel/net/netfilter/xt_connbytes.ko
```
А вы уверены, что не сработает стандартный путь?
```
insmod /lib/modules/$KERNEL/xt_connbytes.ko
```

**Reply by cybsuai** [reply]
Местоположение модуля это конечно в принципе ерунда (сработает ли стандартный путь не знаю, могу проверить если надо, на момент начала ковыряний его включение или не включения не играло роли по причине ниже). 
Главная проблема оказалась в том, что в падавановских прошивках отсутствуют модули xt_multiport и xt_NFQUEUE, и их оказалось необходимо включать в прошивку роутера в момент компиляции. 

А так да, если прошивка уже модифицирована, то ваш оригинальный скрипт работать в принципе должен. Я там чуть чуть и переделал то, просто под своё удобство

**Reply by fakir81** [reply]
Собрал прошивку Padavan с включенными xt_multiport и xt_NFQUEUE под Asus RT-N56U. Nfqws устанавливается, но не стартует. При запуске ошибок нет, а в логе 
initializing conntrack with timeouts tcp=60:300:60 udp=60
nfqws[622]: opening library handle
nfqws[622]: nfq_open(): Protocol not supported
Может кто поможет, что упустил?

**Reply by cybsuai** [reply]
Проверьте также, включен ли у вас CONFIG_IP_NF_QUEUE. Итого должны быть:
CONFIG_IP_NF_QUEUE=y
CONFIG_NETFILTER_XT_TARGET_NFQUEUE=y
CONFIG_NETFILTER_NETLINK_QUEUE=y

**Reply by fakir81** [reply]
Нашел решение  [4pda](https://4pda.to/forum/index.php?showtopic=714487&view=findpost&p=131986938)
В скрипт "Выполнить перед инициализацией маршрутизатора" добавить
modprobe xt_NFQUEUE
modprobe nfnetlink_queue

**Comment by 4Net-Paladin** [comment]
Есть вариант проще. Актуальную прошивку Padavan можно собирать не только с помощью скрипта Prometheus, но и в автоматическом режиме на серверах Github. [Инструкция здесь](https://github.com/shvchk/padavan-builder-workflow/blob/main/README.ru.md).

На 4pda автор этой прошивки и скрипта Prometheus hadzhioglu [подсказал](https://4pda.to/forum/index.php?showtopic=714487&st=19500#entry132528752), что включить NFQUEUE можно не только правкой строки в kernel-3.4.x.config, как это советует делать уважаемый **cybsuai**, но и добавлением в build.config таких строчек:

**# Override kernel config
CONFIG_NETFILTER_XT_TARGET_NFQUEUE=m
CONFIG_NETFILTER_NETLINK_QUEUE=m
CONFIG_IP_NF_QUEUE=m**

А ведь build.config можно (и нужно) редактировать не только в Prometheus, но и [при сборке на серверах Github](https://github.com/shvchk/padavan-builder-workflow/blob/main/README.ru.md). Это гораздо проще, быстрее и доступнее для широких народных масс.

**Reply by cybsuai** [reply]
Думаю сейчас так и надо делать. 

Когда всё начиналось, я же тоже сперва обратился к hadzhioglu с вопросом про добавление, но на тот момент он сказал что таких модулей у нас нет. В итоге сначала у меня получилось добавить их при локальной компиляции, а потом после обсуждения он вроде бы собирался добавить их в свои кастомные прошивки включенными по умолчанию. Респект таким людям!

Ну а пока каждый может себе собрать такой вариант более простым способом.

**Reply by 4Net-Paladin** [reply]
Мне он [написал](https://4pda.to/forum/index.php?showtopic=714487&view=findpost&p=132528632), что, скорее, не будет по умолчанию делать NFQUEUE включенным. Мол, не всем это нужно, а те, кому нужно, добавят три строчки в build.config

Тут, рядом, уже [разместили инструкцию](https://github.com/Anonym-tsk/nfqws-keenetic/discussions/76) для прошивки Padavan - именно с добавлением в build.config этих трех строчек

**Reply by DJ-URAN** [reply]
Собрал по инструкции выше прошивку для mi mini с этими тремя строками в билд конфиге. Обязательно ли после перепрошивки роутера ставить entware? Нельзя ли как-то обойтись без emtware, как это было сделано для tpws?
Также, в инструкции есть пункт установить nfqws по инструкции. А где сама инструкция, как установить-то?

**Reply by votarum** [reply]
> Обязательно ли после перепрошивки роутера ставить entware? Нельзя ли как-то обойтись без emtware, как это было сделано для tpws?
> Также, в инструкции есть пункт установить nfqws по инструкции. А где сама инструкция, как установить-то?

Обязательно. Нельзя. Да на главной же странице она!

**Reply by DJ-URAN** [reply]
А если ставить entware во внутреннюю память роутера, то много ли чего пишет nfgws при работе?

**Reply by votarum** [reply]
На флэш ничего не пишет, в RAM пишет лог, если включите его. С mi mini вам это не грозит, Entware можно поставить только на USB.

**Reply by cybsuai** [reply]
> А если ставить entware во внутреннюю память роутера, то много ли чего пишет nfgws при работе?

На моем ми мини во флеше не было раздела, куда можно поставить entware, но есть usb куда я поставил маленькую флешку и успешно установил всё туда

**Reply by 4Net-Paladin** [reply]
> Собрал по инструкции выше прошивку для mi mini с этими тремя строками в билд конфиге. Обязательно ли после перепрошивки роутера ставить entware? Нельзя ли как-то обойтись без emtware, как это было сделано для tpws? Также, в инструкции есть пункт установить nfqws по инструкции. А где сама инструкция, как установить-то?

Больше подробностей по установке на Padavan прошивки [выложили здесь](https://github.com/Anonym-tsk/nfqws-keenetic/discussions/76). Рекомендую ознакомиться
